_posts/current/blog/2021-01-04-protected-uefi-variables-with-u-boot.md

```diff
@@ -0,0 +1,247 @@
+
+It turns out that the StandAloneMM binary is self-relocatable, so as long as we manage to jump on the first instruction, everything will just 'work'. We would of course need code in OP-TEE to launch the new partition and in U-Boot to communicate with that partition, but that should be way less, or at least that's what we assumed.
+
+And less it was!
+
+[OP-TEE](https://github.com/OP-TEE/optee_os/commit/42471ecf25b7/) and [U-Boot](https://github.com/u-boot/u-boot/commit/f042e47e8fb4/) already got patches for that and EDK2 patches are currently on upstream review.
+
+## **Combining it all together**
+
+So far I've talked about variable management and internal details of Arm's Secure World. We've also managed to run StandAloneMM as part of OP-TEE, but who's responsible for reading and storing the variables eventually?
+
+StandAloneMM includes the driver that implements the accesses to our hardware. EDK2 calls this ***Firmware Volume Block Protocol*** and it's designed to provide control over block-oriented firmware devices. So the missing link is a StandAloneMM FVB that can re-use OP-TEE and it's ability to access our RPMB partition securely, something like [this.](https://git.linaro.org/people/ilias.apalodimas/edk2-platforms.git/tree/Drivers/OpTeeRpmb/OpTeeRpmbFvb.c?h=ffa_svc_optional_on_upstream)
+
+If you combine all of the above, the final architecture looks like this:
+
+{% include image.html path="/assets/images/content/firmware-volume-block-protocol-architecture.png" alt="OP-TEE, U-BOOT & Firmware volume Block Protocol architecture diagram" %}
+An obvious downside is that you need to compile 4 different projects to get your final firmware. On the other hand if we choose to make it a TA (Trusted Application) we use only 3, but you'd have to compile that extra TA and link it to OP-TEE as an EarlyTA anyway.
+
+## Building
+
+### Building TFA
+
+```
+mkdir firmware && cd firmware/
+git clone https://github.com/ARM-software/arm-trusted-firmware
+pushd arm-trusted-firmware/
+make CROSS_COMPILE=aarch64-linux-gnu- ARCH=aarch64 PLAT=<board> TARGET_BOARD=<board> SPD=opteed
+popd
+```
+
+### **Building EDK2**
+
+At the time of this article the EDK2 patchset is under review
+
+```
+git clone https://git.linaro.org/people/ilias.apalodimas/edk2.git -b ffa_svc_optional_on_upstream
+git clone https://git.linaro.org/people/ilias.apalodimas/edk2-platforms.git -b ffa_svc_optional_on_upstream
+export WORKSPACE=$(pwd)
+export PACKAGES_PATH=$WORKSPACE/edk2:$WORKSPACE/edk2-platforms
+export ACTIVE_PLATFORM="Platform/StMMRpmb/PlatformStandaloneMm.dsc"
+export GCC5_AARCH64_PREFIX=aarch64-linux-gnu-
+pushd edk2/
+git submodule init && git submodule update --init --recursive
+popd
+source edk2/edksetup.sh
+make -C edk2/BaseTools
+build -p $ACTIVE_PLATFORM -b RELEASE -a AARCH64 -t GCC5 -n $(nproc)
+```
+
+The StandAloneMM binary is located at **[Build/MmStandaloneRpmb/RELEASE_GCC5/FV/BL32_AP_MM.fd](Build/MmStandaloneRpmb/RELEASE_GCC5/FV/BL32_AP_MM.fd)**
```

Comment:
The StandAloneMM binary is located at **[Build/MmStandaloneRpmb/RELEASE_GCC5/FV/BL32_AP_MM.fd](Build/MmStandaloneRpmb/RELEASE_GCC5/FV/BL32_AP_MM.fd)**

The above is a broken link.

Reply:
Noted with thanks